Minor fixes and enhancements to CollectPoints, TransformProcessor, FiducialRegistrationWizard

tavaughan opened the pull request.

lassoan commented on TransformProcessor/Logic/vtkSlicerTransformProcessorLogic.cxx:
```diff
@@ -441,6 +445,25 @@ void vtkSlicerTransformProcessorLogic::ComputeFullTransform( vtkMRMLTransformPro
+    return;
+  }
+
+  vtkMRMLLinearTransformNode* forwardTransformNode = paramNode->GetInputForwardTransformNode();
```
Use GetTransformBetweenNodes(). It is faster (does not go through world transform, just through the first common parent), works for arbitrary transforms (not just linear), and world transforms (if either or both transforms are NULL).

lassoan commented on TransformProcessor/Logic/vtkSlicerTransformProcessorLogic.cxx:
```diff
@@ -441,6 +445,25 @@ void vtkSlicerTransformProcessorLogic::ComputeFullTransform( vtkMRMLTransformPro
+  vtkMRMLLinearTransformNode* forwardTransformNode = paramNode->GetInputForwardTransformNode();
+  // node stores the transform _to_ parent. Inverse will be the transform _from_ parent.
+  vtkSmartPointer< vtkMatrix4x4 > matrixTransformFromParent = vtkSmartPointer< vtkMatrix4x4 >::New(); 
+  forwardTransformNode->GetMatrixTransformFromParent( matrixTransformFromParent );
```
Is there a way to invert transform between two arbitrary transform nodes? (not just between "forward" and its parent but between "forward" and any other transform)
If you don't have time right now then just add a ticket to keep in mind.

tavaughan replied:
Yes there is, see "Compute Full Transform" mode. You can set a 'from' node and a 'to' node.

lassoan commented on TransformProcessor/qSlicerTransformProcessorModuleWidget.cxx:
```diff
@@ -256,6 +260,7 @@ void qSlicerTransformProcessorModuleWidget::setSignalsBlocked( bool newBlock )
   d->inputInitialTransformComboBox->blockSignals( newBlock );
```
Use qsignalblocker for each widget, we should never overwrite block state without recording the original value (and later restoring the original value).

tavaughan replied:
See new issue #171

lassoan commented on TransformProcessor/qSlicerTransformProcessorModuleWidget.cxx:
```diff
@@ -412,6 +419,8 @@ void qSlicerTransformProcessorModuleWidget::updateInputFieldVisibility()
     d->inputChangedTransformComboBox->setVisible( false );
     d->inputAnchorTransformLabel->setVisible( false );
     d->inputAnchorTransformComboBox->setVisible( false );
+    d->inputForwardTransformLabel->setVisible( false );
```
A lot of code here. It would be much shorter if you called setVisible once for each widget. For example:
d->inputForwardTransformLabel->setVisible(average || shaftPivot)

tavaughan commented:
Thanks Andras for the review, very much appreciated. I think most of your comments can be addressed quickly. I'll create separate issues for the more complex tasks that might involve more discussion (eg, anchor and sampling transforms being or not being the same).

tavaughan commented on CollectPoints/Logic/vtkSlicerCollectPointsLogic.cxx:
```diff
@@ -80,11 +77,14 @@ void vtkSlicerCollectPointsLogic::AddPoint( vtkMRMLCollectPointsNode* collectPoi
+  // find the point coordinates
+  double pointCoordinates[ 3 ] = { 0.0, 0.0, 0.0 }; // temporary values
+  bool success = false; // true if coordinates are computed, false if not.
+  this->ComputePointCoordinates( collectPointsNode, pointCoordinates, success );
```
Yes, I'll do that instead. I wasn't sure which would be nicer style-wise.

tavaughan commented on CollectPoints/Logic/vtkSlicerCollectPointsLogic.cxx:
```diff
@@ -257,7 +303,7 @@ void vtkSlicerCollectPointsLogic::ProcessMRMLNodesEvents( vtkObject* caller, uns
+      if ( collectPointsNode->GetOutputNode() == NULL || collectPointsNode->GetSamplingTransformNode() == NULL )
       {
         vtkWarningMacro( "Collect fiducials node is not fully set up. Setting to manual collection." );
         collectPointsNode->SetCollectModeToManual();
```
I *think* the reason I did this was because the anchor transform is being observed. If the sampling transform is null but the anchor transform is changing rapidly (through OpenIGTLink or another module), *and* we're in auto-collect mode, then this function would be run several times a second. Meaning a warning message is logged several times every second. In hindsight, this might not be a major issue. Should I change this so that it doesn't change the collect mode?

tavaughan replied:
https://github.com/SlicerIGT/SlicerIGT/pull/170/files/ea86a7b8f892cb3befedd5161a26c8a3084cf0c4..1b9a0e0a4521ba12bcbb40916c16420a7436bcf9#r174669150

I wonder now if the anchor transform even needs to be observed. The idea is that the anchor transform provides a basis for points. Should we collect points whenever the anchor moves?

tavaughan commented on CollectPoints/Logic/vtkSlicerCollectPointsLogic.cxx:
```diff
@@ -393,7 +443,14 @@ void vtkSlicerCollectPointsLogic::AddPointToMarkups( vtkMRMLMarkupsFiducialNode*
 
+  // add the label to the point
+  std::stringstream markupLabel;
+  markupLabel << collectPointsNode->GetLabelBase() << collectPointsNode->GetLabelCounter();
```
Is NextLabelCount ok? I want to avoid the word 'Index' because, to me, that suggests it serves a purpose is data storage (like array index).

tavaughan commented on CollectPoints/MRML/vtkMRMLCollectPointsNode.cxx:
```diff
@@ -197,15 +234,15 @@ int vtkMRMLCollectPointsNode::GetNumberOfPointsInOutput()
 {
-  const char* currentNodeId = this->GetNodeReferenceID( OUTPUT_REFERENCE_ROLE );
-  if ( nodeId != NULL && currentNodeId != NULL && strcmp( nodeId, currentNodeId ) == 0 )
+  const char* currentNodeID = this->GetNodeReferenceID( OUTPUT_REFERENCE_ROLE );
```
Are you referring to checking node ID's?

It looks like it still executes a good chunk of code before checking if the node ID is the same or not:
https://github.com/Slicer/Slicer/blob/master/Libs/MRML/Core/vtkMRMLNode.cxx#L1274

lassoan replied:
1. Did you really add that check to improve performance? Probably not. That would have been premature optimization, which should be avoided.

2. Not that execution time would matter that much in this case, but have you checked the chunk of extra code that is executed when you call this->GetNodeReferenceID( OUTPUT_REFERENCE_ROLE )?

3. Most often the code readability and maintainability has the highest priority (over performance, or how easy it is to write the code, etc.). If you remove code then there is less code to read or maintain. That's the main reason I would remove the extra check.

tavaughan replied:
No problem then, I'll remove it. Is there ever a case where we need to check if nodeID == currentNodeID? This is a practice that I've been seeing in SlicerIGT and I've been copying. After I receive a response I'll go through other modules and update accordingly.

tavaughan replied:
(I assume this also applies to your comment "|| (nodeID == NULL && currentNodeID == NULL)"?)

tavaughan commented on CollectPoints/MRML/vtkMRMLCollectPointsNode.cxx:
```diff
@@ -217,7 +254,7 @@ void vtkMRMLCollectPointsNode::ProcessMRMLEvents( vtkObject *caller, unsigned lo
   }
 
-  if ( this->GetProbeTransformNode() && callerNode == this->GetProbeTransformNode() )
+  if ( this->GetSamplingTransformNode() && callerNode == this->GetSamplingTransformNode() )
```
Should just be transform modified, then?

lassoan replied:
Whatever is the event that you added and observer for.

tavaughan replied:
Right. I'll standardize this across the modules I've worked on.

There is sometimes a check for callerNode == NULL. Is this important?

tavaughan commented on CollectPoints/qSlicerCollectPointsModuleWidget.cxx:
```diff
@@ -73,6 +73,21 @@ qSlicerCollectPointsModuleWidget::qSlicerCollectPointsModuleWidget( QWidget* _pa
 qSlicerCollectPointsModuleWidget::~qSlicerCollectPointsModuleWidget()
 {
+  Q_D( qSlicerCollectPointsModuleWidget );
+  disconnect( d->ParameterNodeComboBox, SIGNAL( currentNodeChanged( vtkMRMLNode* ) ), this, SLOT( onParameterNodeSelected() ) );
```
Great! I'll remove these from some of the other modules then.

tavaughan commented on CollectPoints/qSlicerCollectPointsModuleWidget.cxx:
```diff
@@ -147,7 +163,8 @@ void qSlicerCollectPointsModuleWidget::blockAllSignals( bool block )
 {
   Q_D( qSlicerCollectPointsModuleWidget );
   d->ParameterNodeComboBox->blockSignals( block );
```
I can't find documentation for QSignalBlocker in Qt4. Can this wait until we've finished the transition to Qt5 (when Slicer builds on all platforms)?

tavaughan replied:
See new issue #171

tavaughan commented on CollectPoints/Logic/vtkSlicerCollectPointsLogic.cxx:
```diff
@@ -56,22 +56,19 @@ void vtkSlicerCollectPointsLogic::PrintSelf( ostream& os, vtkIndent indent )
-  // find the point coordinates
-  vtkSmartPointer< vtkMRMLLinearTransformNode > probeNode = vtkMRMLLinearTransformNode::SafeDownCast( collectPointsNode->GetProbeTransformNode() );
-  if ( probeNode == NULL )
+  vtkSmartPointer< vtkMRMLLinearTransformNode > samplingNode = vtkMRMLLinearTransformNode::SafeDownCast( collectPointsNode->GetSamplingTransformNode() );
```
I don't see any particular reason why it *has* to be like that. As far as I know, getting a vtkGeneralTransform from GetTransformToNode() should work.

tavaughan replied:
Is there any particular application in mind for non-linear transforms?

lassoan replied:
We may have a warping transform in the transformation chain (e.g., from Fiducial registration wizard).

tavaughan commented on CollectPoints/Logic/vtkSlicerCollectPointsLogic.cxx:
```diff
@@ -114,6 +108,58 @@ void vtkSlicerCollectPointsLogic::AddPoint( vtkMRMLCollectPointsNode* collectPoi
+  }
+
+  // find the point coordinates
+  vtkSmartPointer< vtkMRMLLinearTransformNode > samplingNode = vtkMRMLLinearTransformNode::SafeDownCast( collectPointsNode->GetSamplingTransformNode() );
```
They seem to, yes.

lassoan replied:
Great, then instead of vtkMRMLLinearTransformNode, let's use vtkTransformNode everywhere.

tavaughan commented on CollectPoints/qSlicerCollectPointsModuleWidget.cxx:
```diff
@@ -263,26 +263,17 @@ void qSlicerCollectPointsModuleWidget::onAnchorTransformNodeSelected()
   }
+  collectPointsNode->CreateDefaultDisplayNodesForOutputNode();
 
   this->updateGUIFromMRML();
```
I assume you refer to updateGUIFromMRML? I always do this just in case something at the node level necessitates change in some of the attributes/references.

It also gives a cue when something in the node isn't updating the way it is supposed to (say I choose an output node and for some reason it isn't actually set, then I'll see this in the GUI right away. I won't need to assume that the node I see in the box has been set successfully, I will know for sure one way or the other.)

lassoan replied:
>  I always do this just in case something at the node level necessitates change in some of the attributes/references.

Node changes are already observed (if not, then it's an error that has to be fixed). So, any extra updateGUIFromMRML calls are unnecessary. The computation time is negligible, but the main issue is that it may mask the error of missing node observations.

tavaughan replied:
I see. I think most of the modules I've worked with have called updateGUIFromMRML (or the equivalent) manually. I'd been following those examples. What is the proper mechanism for updating the GUI?

lassoan replied:
When you select a node, you add an observer to modification events: 

https://github.com/SlicerIGT/SlicerIGT/blob/8d7fbdc08d39ff6a14f9f38746c0fce7c88444e9/ToolWatchdog/qSlicerWatchdogModuleWidget.cxx#L226-L233

https://github.com/SlicerIGT/SlicerIGT/blob/8d7fbdc08d39ff6a14f9f38746c0fce7c88444e9/FiducialRegistrationWizard/qSlicerFiducialRegistrationWizardModuleWidget.cxx#L488-L495

tavaughan replied:
Thanks, I will make sure that the modules I've worked with recently follow those two examples.

lassoan commented:
> I wonder now if the anchor transform even needs to be observed. 

YES! It is very important to observe the anchor node transform modification changes as well. If we find that performance may suffer then we can limit the maximum update rate, but we don't need to worry about it now.

lassoan commented:
> I think the reason I did this was because the anchor transform is being observed. If the sampling transform is null but the anchor transform is changing rapidly (through OpenIGTLink or another module), and we're in auto-collect mode, then this function would be run several times a second. Meaning a warning message is logged several times every second. In hindsight, this might not be a major issue. Should I change this so that it doesn't change the collect mode?

I see. We should definitely not change the collection mode. NULL sampling transform is an error (you sample the world coordinate system), there should be no warning.
If output is NULL then you may log a warning, but probably it is enough if you indicate the error at the GUI level for the user. I would not like to see warning messages in the log if I select input transforms before creating/selecting the output node.

tavaughan commented:
I think I have addressed all of the issues.

> If output is NULL then you may log a warning, but probably it is enough if you indicate the error at the GUI level for the user.

Perhaps this is something we should standardize across modules. I like how fiducial registration wizard shows status messages in the GUI. But I think almost all the other modules log warnings or errors. Should this be its own ticket?
